txpipe/twopoint_null_tests.py

```diff
@@ -766,3 +767,272 @@ def write_output(self, source_list, lens_list, meta, results):
+                nbin_source = f.file["maps"].attrs["nbin_source"]
+                g1_maps     = [f.read_map(f"g1_{b}") for b in range(nbin_source)]
+                g2_maps     = [f.read_map(f"g2_{b}") for b in range(nbin_source)]
+                weights     = [f.read_map(f"lensing_weight_{b}") for b in range(nbin_source)]
+
+            # Open mask
+            with self.open_input("mask", wrapper=True) as f:
+                mask = f.read_map("mask")
+            
+            # Define binning 
+            b     = nmt.NmtBin.from_nside_linear(hp.npix2nside(mask.shape[0]), self.config["Nell"])
+            
+            # Initialize the fields
+            fields = {}
+            for i in range(nbin_source):
+                g1_maps[i][g1_maps[i]==hp.UNSEEN]=0
+                g2_maps[i][g2_maps[i]==hp.UNSEEN]=0
+                fields[i] = nmt.NmtField(mask, [g1_maps[i], g2_maps[i]], purify_e=False, purify_b=purify_b)
+
+            # Compute Cls and store them in dictionrary
+            tmp       = np.array([])
+            ret=np.zeros((nbin_source,nbin_source,len(b.get_effective_ells())))
+
+            for zi in range(nbin_source):
+                for zj in range(zi,nbin_source):
+                     
+                    field1 = fields[zi]
+                    field2 = fields[zj]
+                    w_yp  = nmt.NmtWorkspace()
+                    w_yp.compute_coupling_matrix(field1, field2, b)
+    
+                    cl_coupled   = nmt.compute_coupled_cell(field1,field2)
+                    cl_decoupled = w_yp.decouple_cell(cl_coupled)
+                    ret[zj,zi,:]= cl_decoupled[3]
+                    tmp = np.concatenate([tmp,cl_decoupled[3]])
+
+            
+            # Compute covariance in a somewhat adhoc way by shuffling the pixel values.
+            covarr = np.zeros((len(tmp),self.config['Nsims']))
+
+            for k in tqdm(range(self.config['Nsims'])):
+                fields = {}
+                for i in range(nbin_source):
+                    idx1 = np.where(g1_maps[i]!=0)[0]
+                    idx2 = np.random.permutation(idx1)
+                    g1_maps[i][idx1]=g1_maps[i][idx2]
+                    g2_maps[i][idx1]=g2_maps[i][idx2]
+                    fields[i] = nmt.NmtField(mask, [g1_maps[i], g2_maps[i]], purify_e=False, purify_b=purify_b)
+
+                tmp = np.array([])
+                for zi in range(nbin_source):
+                    for zj in range(zi,nbin_source):
+                        
+                        field1 = fields[zi]
+                        field2 = fields[zj]
+                        w_yp  = nmt.NmtWorkspace()
+                        w_yp.compute_coupling_matrix(field1, field2, b)
+        
+                        cl_coupled   = nmt.compute_coupled_cell(field1,field2)
+                        cl_decoupled = w_yp.decouple_cell(cl_coupled)
+                        tmp = np.concatenate([tmp,cl_decoupled[3]])
```

Comment:
Not quite clear what this tmp array is for. It's only used to get the size in a moment, but then you can just accumulate that.

Reply:
Decided to directly get the array size with the known number of ell bins and numbers of zbin combination so this is not needed anymore. Removed.

Reply:
This `tmp` is slightly different from the comment above. We need to estimate the full covariance that includes all the redshift combinations, which means we must create a matrix with each column being the full concatenated data vector and the different columns representing different realizations (to which we  apply np.cov later)